### PR TITLE
[WIP] Adding Baseline to Reinforce

### PR DESCRIPTION
**Pull request by psjanani**
Currently, the Regressor is trained at sentence level as the process is very slow otherwise.

### CODE REVIEW

File: xnmt/translator.py [hunk 1]
```diff
@@ -236,12 +237,33 @@ def __call__(self, translator, dec_state, src, trg):
...
+      self.param_collection = dy.Model()
+      self.trainer = dy.AdamTrainer(self.param_collection)
+      self.pW = self.param_collection.add_parameters((1, dim))
+      self.pb = self.param_collection.add_parameters(1)
+
+    def predict_score(self, h_t, dim):
+      h_t = dy.inputTensor(np.reshape(h_t, (dim[0][0], dim[1])), batched=True)
+      W = dy.parameter(self.pW)
+      B = dy.parameter(self.pb)
+      output = B + W * h_t
```
Comment (philip30) [inline xnmt/translator.py]: Use linear.Linear instead

File: xnmt/translator.py [hunk 2]
```diff
@@ -236,12 +237,33 @@ def __call__(self, translator, dec_state, src, trg):
 class TranslatorReinforceLoss(Serializable):
   yaml_tag = '!TranslatorReinforceLoss'
 
+  class Regressor():
+    def __init__(self, dim):
+      self.param_collection = dy.Model()
+      self.trainer = dy.AdamTrainer(self.param_collection)
```
Comment (philip30) [inline xnmt/translator.py]: Not sure if adding 1 more trainer here is the correct way to incorporate the baseline. See https://github.com/neulab/xnmt/blob/segmenting_encoder/xnmt/segmenting_encoder.py#L211 for example. The idea is to compose the loss of the decoder and baseline together.

File: xnmt/translator.py [hunk 3]
```diff
@@ -236,12 +237,33 @@ def __call__(self, translator, dec_state, src, trg):
...
+      self.pW = self.param_collection.add_parameters((1, dim))
+      self.pb = self.param_collection.add_parameters(1)
+
+    def predict_score(self, h_t, dim):
+      h_t = dy.inputTensor(np.reshape(h_t, (dim[0][0], dim[1])), batched=True)
```
Comment (philip30) [inline xnmt/translator.py]: use dy.nobackprop() instead to create a copy & non backpropagate-able graph

File: xnmt/translator.py [hunk 4]
```diff
@@ -236,12 +237,33 @@ def __call__(self, translator, dec_state, src, trg):
 class TranslatorReinforceLoss(Serializable):
   yaml_tag = '!TranslatorReinforceLoss'
 
+  class Regressor():
```
Comment (philip30) [inline xnmt/translator.py]: Regressor can be implemented by a single Linear Transformation object located in linear.Linear.

File: xnmt/translator.py [hunk 5]
```diff
@@ -277,9 +307,16 @@ def __call__(self, translator, dec_state, src, trg):
       except ValueError:
         pass
       # Calculate the evaluation score
-      eval_score.append(self.evaluation_metric.evaluate_fast(trg_i.words, sample_i))
+      score = self.evaluation_metric.evaluate_fast(trg_i.words, sample_i)
+      eval_score.append(0 if math.isnan(score) else score)
```
Comment (philip30) [inline xnmt/translator.py]: Do not squash nan to 0, this may create a more severe bug in the future I think.

File: xnmt/translator.py [hunk 6]
```diff
@@ -277,9 +307,16 @@ def __call__(self, translator, dec_state, src, trg):
...
+
+    true = dy.inputTensor(eval_score, batched=True)
+    loss_exp = self.regressor.calc_loss(decode_hidden_state, eval_score, dim)
+    loss_exp.backward()
+    self.regressor.trainer.update()
```
Comment (philip30) [inline xnmt/translator.py]: There shouldn't be any training update or backward in calc_loss function

Comment (psjanani) [reply xnmt/translator.py]: Oh, I thought this would train the Regressor according to the inputs. How do you suggest that t=should be done?

File: xnmt/translator.py (hunk 5 above)
Comment (psjanani) [reply xnmt/translator.py]: Makes sense. Currently, this nan is thrown when the target sentence is of zero length. I can test that before.

File: xnmt/translator.py (hunk 6 above)
Comment (psjanani) [reply xnmt/translator.py]: Okay, got it. I shall look into the segmenting Encoder as suggested by you. Thanks.

File: xnmt/translator.py [hunk 7]
```diff
@@ -277,9 +292,23 @@ def __call__(self, translator, dec_state, src, trg):
...
+      self.eval_score.append(score)
+      self.trueScore = dy.inputTensor(self.eval_score, batched=True)
+
+    final_score = -self.trueScore
+    if self.use_baseline:
+      final_score = self.bs - self.trueScore
+    return dy.sum_elems(dy.cmult(final_score , dy.esum(logsofts)))
+
+  @recursive_sum
+  def calc_additional_loss(self, _):
```
Comment (philip30) [inline xnmt/translator.py]: I don't think you need to do this. This is required if you are calculating the reward. In this case, the loss can be composed inside the __call__ and the LossBuilder() object is returned instead.

File: xnmt/translator.py [hunk 8]
```diff
@@ -262,8 +269,16 @@ def __call__(self, translator, dec_state, src, trg):
       done = list(six.moves.map(lambda x: x == Vocab.ES, sample))
       if all(done):
         break
+
+    if self.use_baseline:
+      h_t = dy.tanh(translator.decoder.context_projector(dy.concatenate([dec_state.rnn_state.output(), dec_state.context])))
+      if not self.baseline:
+        model = translator.decoder.param_col
```
Comment (philip30) [inline xnmt/translator.py]: Initiate this in the constructor

File: xnmt/translator.py [hunk 9]
```diff
@@ -277,9 +292,23 @@ def __call__(self, translator, dec_state, src, trg):
...
       # Calculate the evaluation score
-      eval_score.append(self.evaluation_metric.evaluate_fast(trg_i.words, sample_i))
-
-    return -dy.sum_elems(dy.cmult(dy.inputTensor(eval_score, batched=True), dy.esum(logsofts)))
+      score = 0 if not len(sample_i) else self.evaluation_metric.evaluate_fast(trg_i.words, sample_i)
+      self.eval_score.append(score)
+      self.trueScore = dy.inputTensor(self.eval_score, batched=True)
+
+    final_score = -self.trueScore
+    if self.use_baseline:
```
Comment (philip30) [inline xnmt/translator.py]: I think this is a bit different than I know. The baseline should be predicted at every timestep and be multiplied with the log prob at that time step, right? @neubig

File: xnmt/translator.py (hunk 8 above)
Comment (philip30) [reply xnmt/translator.py]: Find a better way to pass the information about the size of the input. But this should be done in the constructor I think.

File: xnmt/translator.py [hunk 10]
```diff
@@ -74,7 +78,8 @@ def __init__(self, src_embedder, encoder, attender, trg_embedder, decoder, loss_
       self.loss_calculator = TranslatorMLELoss()
     else:
       self.loss_calculator = loss_calculator
-
+    if (issubclass(self.loss_calculator.__class__, HierarchicalModel)):
+      self.register_hier_child(self.loss_calculator)
```
Comment (philip30) [inline xnmt/translator.py]: No need to check this. We can assume that the subclass is a HierarchicalModel

File: xnmt/translator.py (hunk 9 above)
Comment (psjanani) [reply xnmt/translator.py]: Oh sorry, I initially read the algorithm and assumed that the word level predictions would be time-consuming. (I thought that the loss for Linear Regression would be according to BLEU score at every t-1th step which seemed too costly and ambiguous). 
So the intention was to develop the pipeline for sentence level approximation and then get to the word level in the next commit. But it seems to require just overall BLEU Score which makes complete sense and is not slow. Shall change it in next iteration. Thanks.

File: xnmt/translator.py (hunk 7 above)
Comment (psjanani) [reply xnmt/translator.py]: Oh, I am confused here, @philip30. Are you suggesting to make the standard-loss an instance of LossBuilder? or am I misunderstanding something? Currently, I guess xnmt_train the "standard loss" (from the model's calc_loss) is considered by default to be a single loss and all additional losses seem to go as an instance of LossBuilder.  Wouldn't further generalizing this format make things more confusing?

File: xnmt/translator.py (hunk 9 above)
Comment (philip30) [reply xnmt/translator.py]: Ah yes.. You are correct. I think this is true if we are using the reward for the whole sequence :) Please pardon my previous comment.

File: xnmt/translator.py (hunk 7 above)
Comment (philip30) [reply xnmt/translator.py]: You can do this to substitute the return in __call__:
``` python
loss = LossBuilder()
loss.add_loss("reinf_loss", dy.sum_elems(dy.cmult(final_score , dy.esum(logsofts))))
loss.add_loss("baseline", dy.sum_batches(dy.squared_distance(self.trueScore, self.bs))
return loss
```
It doesn't matter if you return a loss_builder or an instance of dynet_expression as far as I know.

Comment (psjanani) [reply xnmt/translator.py]: Hi thanks a lot for you quick replies.
When I do this, I get the following error:
```     
   File "/xnmt/xnmt/xnmt_train.py", line 231, in run_epoch
   additional_loss = self.model.calc_additional_loss(dy.nobackprop(-standard_loss))   
   TypeError   :    bad operand type for unary -: 'LossBuilder'  
```

So, it seems that main algorithm doesn't expect LossBuilder object there. I can change it. But I wanted to ask if it is a good design choice.

Comment (philip30) [reply xnmt/translator.py]: I see. This is a new case, but I think the calc_loss should handle both the LossBuilder and dynet expression object. You can fix it with a design you think is good.
I think you can either implement the unary operator in LossBuilder or add a conditional statment there.

Comment (philip30) [reply xnmt/translator.py]: Or if it is too difficult, you can neglect the LossBuilder. Instead just use the "+" to sum 2 dynet expression. LossBuilder is used only for reporting purposes (so we can see 2 losses are going down separately)

File: xnmt/translator.py (hunk 9 above)
Comment (psjanani) [reply xnmt/translator.py]: Hi!  I guess your initial observation seems to be valid. I might be wrong, but by adding the projected score (the projection of the final BLEU score) at every instance, we might be able to penalize the sentences that deviated early on, more.
